examples/vision/metric_learning.py

```diff
@@ -22,290 +24,380 @@
+check that it contains the expected number of classes and examples per class.
 """
 
-num_classes = 10
+x_batch, y_batch = train_ds.generate_batch(0)
+unique_class_ids, _, class_counts = tf.unique_with_counts(y_batch)
+print(f'The batch contains {len(x_batch)} examples in total\n')
+for cid, cc in zip(unique_class_ids, class_counts):
+    print(f'Class ID {cid} has {cc} examples in the batch')
 
+"""
+# Embedding Model
 
-class AnchorPositivePairs(keras.utils.Sequence):
-    def __init__(self, num_batchs):
-        self.num_batchs = num_batchs
+Next we define a `SimilarityModel()` using the functional Keras API. The model
+is a standard ConvNet with the addition of a `MetricEmbedding()` layer that
+applies L2 normalization. The metric embedding layer is helpful when using
+`Cosine` distance as we only care about the angle between the vectors.
 
-    def __len__(self):
-        return self.num_batchs
+Additionally, the `SimilarityModel()` provides a number of helper methods for:
 
-    def __getitem__(self, _idx):
-        x = np.empty((2, num_classes, height_width, height_width, 3), dtype=np.float32)
-        for class_idx in range(num_classes):
-            examples_for_class = class_idx_to_train_idxs[class_idx]
-            anchor_idx = random.choice(examples_for_class)
-            positive_idx = random.choice(examples_for_class)
-            while positive_idx == anchor_idx:
-                positive_idx = random.choice(examples_for_class)
-            x[0, class_idx] = x_train[anchor_idx]
-            x[1, class_idx] = x_train[positive_idx]
-        return x
+* indexing embedded examples
+* performing example lookups
+* evaluating the classification
+* evaluating the quality of the embedding space
+"""
 
+embedding_size = 256
+
+inputs = layers.Input((32, 32, 3))
+x = inputs
+x = layers.Conv2D(64, 3, activation='relu')(x)
+x = layers.BatchNormalization()(x)
+x = layers.Conv2D(128, 3, activation='relu')(x)
+x = layers.BatchNormalization()(x)
+x = layers.MaxPool2D((4, 4))(x)
+x = layers.Conv2D(256, 3, activation='relu')(x)
+x = layers.BatchNormalization()(x)
+x = layers.Conv2D(256, 3, activation='relu')(x)
+x = layers.GlobalMaxPool2D()(x)
+outputs = MetricEmbedding(embedding_size)(x)
+
+# building model
+model = SimilarityModel(inputs, outputs)
+model.summary()
 
 """
-We can visualise a batch in another collage. The top row shows randomly chosen anchors
-from the 10 classes, the bottom row shows the corresponding 10 positives.
+## Similarity Loss
+
+The similarity loss expects batches containing at least 2 examples of each
+class, from which it computes the loss over the pairwise positive and negative
+distances. Here we are using `MultiSimilarityLoss()`, one of several losses in
+Tensorflow Similarity. This loss attempts to use all informative pairs in the
+batch, taking into account the self-similarity, positive-similarity, and the
+negative-similarity.
+
+For more information on the MS loss see
```

Comment:
Remove and simply link the paper next to the MultiSimilariyt() part.

Reply:
Done